pkg/planner/core/expression_rewriter.go

```diff
@@ -2640,6 +2641,185 @@ func hasCurrentDatetimeDefault(col *model.ColumnInfo) bool {
+		if err != nil {
+			return nil, false, errors.Trace(err)
+		}
+		tbl = part.GetPartition(pid)
+	} else {
+		if len(partName) != 0 {
+			return nil, false, errors.New("not partition table")
+		}
+	}
+	recordID, err := buildHandle(ctx, tbl.Meta(), args[2:], row)
+	if err != nil {
+		return nil, false, err
+	}
+	key := tablecodec.EncodeRecordKey(tbl.RecordPrefix(), recordID)
+	return key, false, nil
+}
+
+func extractTablePartition(str string) (table, partition string) {
+	start := strings.IndexByte(str, '(')
+	if start == -1 {
+		return str, ""
+	}
+	end := strings.IndexByte(str, ')')
+	if end == -1 {
+		return str, ""
+	}
+	return str[:start], str[start+1 : end]
+}
+
+func buildHandle(
+	ctx expression.EvalContext,
+	tblInfo *model.TableInfo,
+	pkArgs []expression.Expression,
+	row chunk.Row,
+) (kv.Handle, error) {
+	var recordID kv.Handle
+	if !tblInfo.IsCommonHandle {
+		h, _, err := pkArgs[0].EvalInt(ctx, row)
+		if err != nil {
+			return nil, err
+		}
+		recordID = kv.IntHandle(h)
+	} else {
+		pkIdx := tables.FindPrimaryIndex(tblInfo)
+		if len(pkIdx.Columns) != len(pkArgs) {
+			return nil, errors.Errorf("pk column count mismatch, expected %d, got %d", len(pkIdx.Columns), pkArgs)
+		}
+		pkDts := make([]types.Datum, 0, len(pkIdx.Columns))
+		for i, idxCol := range pkIdx.Columns {
+			dt, err := pkArgs[i].Eval(ctx, row)
+			if err != nil {
+				return nil, err
+			}
+			ft := tblInfo.Columns[idxCol.Offset].FieldType
+			pkDt, err := dt.ConvertTo(ctx.TypeCtx(), &ft)
+			if err != nil {
+				return nil, err
+			}
+			pkDts = append(pkDts, pkDt)
+		}
+		tablecodec.TruncateIndexValues(tblInfo, pkIdx, pkDts)
+		var handleBytes []byte
+		handleBytes, err := codec.EncodeKey(ctx.Location(), nil, pkDts...)
+		ec := ctx.ErrCtx()
+		err = ec.HandleError(err)
+		if err != nil {
+			return nil, err
+		}
+		recordID, err = kv.NewCommonHandle(handleBytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return recordID, nil
+}
+
+func encodeIndexKeyFromRow(
+	ctx expression.EvalContext,
+	isVer infoschemactx.MetaOnlyInfoSchema,
+	args []expression.Expression,
+	row chunk.Row,
+) ([]byte, bool, error) {
+	dbName, isNull, err := args[0].EvalString(ctx, row)
+	if err != nil || isNull {
+		return nil, isNull, err
+	}
+	tblName, isNull, err := args[1].EvalString(ctx, row)
+	if err != nil || isNull {
+		return nil, isNull, err
+	}
+	if !ctx.RequestVerification(dbName, tblName, "", mysql.AllPrivMask) {
+		// The arguments will be filled by caller.
+		return nil, false, plannererrors.ErrSpecificAccessDenied
+	}
+	idxName, isNull, err := args[2].EvalString(ctx, row)
+	if err != nil || isNull {
+		return nil, isNull, err
+	}
+	is := isVer.(infoschema.InfoSchema)
+	if is == nil {
+		return nil, false, errors.New("missing information schema")
+	}
+	tbl, err := is.TableByName(context.Background(), model.NewCIStr(dbName), model.NewCIStr(tblName))
+	if err != nil {
+		return nil, false, err
+	}
+	tblInfo := tbl.Meta()
```

Comment:
is `tbl` actually needed here, or is `TableInfoByName` sufficient (TableInfoByName requires only MetaOnlyInfoSchema).

Reply:
Yes, we need `table.Table` to choose the correct partition. Previous implementation does not consider partition. This bug is fixed in https://github.com/pingcap/tidb/pull/51678/commits/4e07b96d212ad91e4ca8e1d272c9147a72c8b18b.